### PR TITLE
MG-157: Integrate must-gather-clean in must-gather-operator.

### PR DESCRIPTION
**Pull request by shivprakashmuley**
## Summary by CodeRabbit

* **New Features**
  * Added `spec.obfuscate` to the MustGather CRD for obfuscation-and-store/upload, obfuscate-only from an existing bundle, and obfuscate + upload.
  * Introduced configurable obfuscation rules via a referenced ConfigMap and PVC-backed “source” mode, with expanded CRD validation for supported option combinations.
  * Obfuscation is applied during MustGather jobs and uses an obfuscation-aware upload flow; the referenced obfuscation ConfigMap is automatically managed for the run.
* **Bug Fixes**
  * When obfuscation is enabled, uploads skip gracefully (no failure) if SFTP credentials are missing.
* **Documentation**
  * Added new example manifests covering the new obfuscation modes.

### CODE REVIEW

**Comment by Prashanth684** [comment]
---
  Deep Review Verdict
  
  Disposition: REQUEST_CHANGES — 5 blocking findings, 3 with confirmed reproducers

  ---
  Specialist Findings

  Bugs (3 findings):
  1. BLOCKING — Double-nested cleaned/cleaned/ directory in Gather+Obfuscate+PVC mode (template.go:431 + upload:76) — reproduced
  2. BLOCKING — Obfuscate-only source mode silently discards all cleaned output to ephemeral emptyDir (template.go:234) — reproduced
  3. SUGGESTION — Custom gather command + obfuscation skips chown, causing permission errors (template.go:350)

  Adversarial (16 findings, 3 unique after dedup):
  1. BLOCKING — chown -R 65534:65534 fails for non-root must-gather images, causing 3 retries then failure (constant.go:37) — reproduced
  2. SUGGESTION — No CRD validation prevents obfuscate.source + gatherSpec/imageStreamRef — gather fields silently ignored, ImageStream resolution can block obfuscation
  3. SUGGESTION — Spec changed from required to optional allows creation of empty MustGather CRs that produce no output
  4. Corroborated findings #1, #2 from Bugs specialist and multiple findings below

  Security & Supply Chain (2 blocking, 3 low):
  1. BLOCKING — CSV manifest contains personal dev image quay.io/rh-ee-smuley/mg-op-test:july23261 — supply chain risk (support-log-gather-operator.clusterserviceversion.yaml:230,249)
  2. BLOCKING — Security-critical golang.org/x/ module version downgrades: x/crypto v0.53→v0.47, x/net v0.55→v0.48, x/sys v0.46→v0.40, x/text v0.38→v0.33 — may reintroduce CVEs (go.mod)
  3. NOTE — No subPath path-traversal validation (defense-in-depth, kubelet validates at runtime)

  Architecture (12 findings, key unique ones):
  1. BLOCKING — CRD YAML manually edited to diverge from Go type markers for serviceAccountName (Required+MinLength=1 in Go, optional+default in YAML) and Spec (Required in Go, omitempty in YAML). Running make manifests will revert YAML changes. (mustgather_types.go:287-293 + CRD YAMLs)
  2. SUGGESTION — Shell/Go path coupling: obfuscation output path cleaned hardcoded in both template.go:431 and upload:76 with no shared constant (root cause of Bug #1)
  3. SUGGESTION — Custom command + obfuscation silently skips chown with no CRD validation preventing the combination
  4. SUGGESTION — No validation prevents obfuscate.source + storage, producing undocumented dual-PVC semantics
  5. NOTE — getUploadContainer has 9 parameters, growing function complexity

  Codebase Consistency (8 findings):
  1. SUGGESTION — ensureObfuscateConfigMap/cleanupObfuscateConfigMap are near-verbatim copies of the TrustedCA equivalents (~140 lines). Extract a generic helper.
  2. SUGGESTION — Duplicate helpers: obfuscateConfigMapRefName (controller) vs obfuscateConfigMapName (template) do the same nil-safe extraction 
  3. SUGGESTION — Verbose triple-nested nil-check at lines 178 and 535 should use existing obfuscateConfigMapRefName helper
  4. NOTE — Missing Apache 2.0 copyright headers on new pkg/mustgatherutil/ files
  5. NOTE — Unused constant obfuscateConfigMountDir (dead code)

  QA Engineer (16 findings):
  1. SUGGESTION — ensureObfuscateConfigMap and cleanupObfuscateConfigMap have zero test coverage despite identical complexity to the fully-tested TrustedCA equivalents
  2. SUGGESTION — No test for the most common mode: Gather+Obfuscate+Upload (SFTP) 
  3. SUGGESTION — No test for Obfuscate+Upload from Source mode
  4. SUGGESTION — sftpUploadParams, initializeJobTemplate with obfuscation, and several edge cases untested

  Technical Writer (6 findings):
  1. SUGGESTION — README not updated for the new obfuscation feature (4 operational modes, new API fields)
  2. NOTE — CLAUDE.md sections (API Types, Job Template, Upload Script, Reconciliation Flow) are stale after this change
  3. NOTE — Mode numbering in API godoc doesn't match example file labels

  ---
  Panel Synthesis
  
  Six specialists independently flagged the personal test image in the CSV — this is the strongest corroboration in the review. The image quay.io/rh-ee-smuley/mg-op-test:july23261 is clearly a development artifact that must be reverted.

  The two highest-impact bugs — double-nested cleaned/cleaned/ directory and obfuscate-only data loss — were found independently by both the Bugs and Adversarial specialists and confirmed through code analysis. Both are logic errors in template.go's volume/mount setup that will cause silent data corruption or loss
  in production:

  - Double-nested cleaned: The fix is to either remove "cleaned" from the SubPath in template.go:431 (letting the shell script create it) OR remove the mkdir -p /must-gather-upload/cleaned from the upload script and pass the mount point directly. The shell script should be the single source of truth for the output
  directory name.
  - Obfuscate-only data loss: The CRD validation rule at mustgather_types.go:30 accepts obfuscate.source alone, but the template code at line 234 excludes source mode from PVC-backed upload volumes. Either the CRD rule should require storage or uploadTarget alongside source, or the template should handle this mode
  by persisting output somewhere accessible.

  The dependency downgrade (Security) is a supply-chain concern: must-gather-clean v0.0.5 pulls older golang.org/x/ modules. A govulncheck run should determine if this introduces known CVEs. If so, must-gather-clean needs to be updated or the go.mod needs explicit require directives to pin higher versions.

  The CRD/Go type drift was caught by Architecture, Consistency, and Adversarial reviewers. The Go markers and the shipped CRD disagree on whether serviceAccountName is required. This is a ticking time bomb — the next make manifests run will silently revert the YAML changes.

  The chown non-root failure (Adversarial) is a real bug but narrower in scope — it only affects custom must-gather images that run as non-root, which is uncommon but valid. A simple mitigation would be to add || true after the chown or use a wrapper that handles the error gracefully.

  The Consistency and QA reviewers agreed that the duplicated ConfigMap lifecycle code (~140 lines) should be refactored into a generic helper. This is not blocking but is worth addressing to reduce maintenance burden.

  ---
  Required Actions Before Merge
  
  1. Revert personal test image in bundle/manifests/tech-preview/support-log-gather-operator.clusterserviceversion.yaml:230,249 — replace quay.io/rh-ee-smuley/mg-op-test:july23261 with the production image
  2. Fix double-nested cleaned/cleaned/ directory — reconcile the SubPath in template.go:431 with the output path in upload:76 so obfuscation output lands at the correct PVC path
  3. Fix obfuscate-only source mode data loss — either require storage or uploadTarget when obfuscate.source is used (tighten CRD validation), or back the upload volume with a user-specified PVC in this mode
  4. Reconcile CRD YAML with Go type markers for serviceAccountName and Spec — update the Go markers to be the source of truth (api/v1alpha1/mustgather_types.go:287-293) and regenerate CRDs with make manifests
  5. Verify dependency downgrades are safe — run govulncheck against the downgraded golang.org/x/crypto v0.47 and golang.org/x/net v0.48. If CVEs exist, pin higher versions in go.mod or update must-gather-clean

  ---
  Optional Follow-ups
  
  - Handle chown failure for non-root must-gather images (add || true, CAP_CHOWN, or CRD validation)
  - Add CRD validation preventing obfuscate.source + gatherSpec.command/imageStreamRef (silently ignored fields)
  - Extract generic ensureCrossNamespaceConfigMap/cleanupCrossNamespaceConfigMap helper to eliminate ~140 lines of duplication
  - Add unit tests for ensureObfuscateConfigMap/cleanupObfuscateConfigMap (mirror trusted_ca_test.go)
  - Add tests for Gather+Obfuscate+Upload and Obfuscate+Upload+Source modes
  - Add Apache 2.0 copyright headers to pkg/mustgatherutil/ files
  - Remove unused constant obfuscateConfigMountDir
  - Update README with obfuscation documentation
  - Use obfuscateConfigMapRefName(instance) != "" instead of inline triple-nil-check

  ---
  Stats
  
  Arbiter summary: 48 raw findings from 7 specialists.
  Kept: 5 blocking (3 reproduced), 12 suggestions/notes.
  Dropped: 31 duplicates or corroborations merged into primary findings.

  Generated by [/deep-review](https://github.com/stbenjam/claude-nine/blob/main/plugins/reviews/skills/deep-review/SKILL.md)

File: api/v1alpha1/mustgather_types.go [hunk 1]
```diff
@@ -207,6 +223,49 @@ type PersistentVolumeClaimReference struct {
...
+	// consistently replaces IPs and MACs, and omits Secrets and ConfigMaps.
+	// +optional
+	ObfuscationConfigRef *corev1.LocalObjectReference `json:"obfuscationConfigRef,omitempty"`
+
+	// source references an existing must-gather bundle on a PVC
+	// for obfuscation without running a new gather.
+	// When set, the operator skips the gather step and runs obfuscation
+	// directly on the referenced PVC contents.
+	// +optional
+	Source *ObfuscateSourceConfig `json:"source,omitempty"`
```
Comment (swghosh) [inline api/v1alpha1/mustgather_types.go]: please reuse the `PersistentVolumeConfig` here directly to avoid any duplicates in field values.

Apparently, both `PersistentVolumeConfig` and `ObfuscateSourceConfig` have the same 2 fields and serve the same purpose altogether.

File: controllers/mustgather/constant.go [hunk 2]
```diff
@@ -22,4 +22,18 @@ const (
 
 	// DefaultMustGatherImageEnv represents the environment variable for the default must-gather image
 	DefaultMustGatherImageEnv = "DEFAULT_MUST_GATHER_IMAGE"
+
+	// Obfuscation env vars consumed by build/bin/upload.
+	obfuscateEnvEnabled = "obfuscate"
```
Comment (swghosh) [inline controllers/mustgather/constant.go:27]:
```suggestion
	obfuscateEnvEnabled = "OBFUSCATE_ENABLED"
```

File: controllers/mustgather/constant.go [hunk 3]
```diff
@@ -22,4 +22,18 @@ const (
...
+	obfuscateEnvConfig  = "obfuscate_config"
```
Comment (swghosh) [inline controllers/mustgather/constant.go:28]:
```suggestion
	obfuscateEnvConfig  = "OBFUSCATE_CONFIG_PATH"
```

File: controllers/mustgather/mustgather_controller.go [hunk 4]
```diff
@@ -674,3 +688,149 @@ func (r *MustGatherReconciler) cleanupTrustedCAConfigMap(ctx context.Context, re
...
+func obfuscateConfigMapRefName(instance *mustgatherv1alpha1.MustGather) string {
+	if instance.Spec.Obfuscate == nil || instance.Spec.Obfuscate.ObfuscationConfigRef == nil {
+		return ""
+	}
+	return instance.Spec.Obfuscate.ObfuscationConfigRef.Name
+}
+
+// ensureObfuscateConfigMap copies the obfuscation config ConfigMap from the operator namespace
+// to the CR namespace when they differ, and adds/updates the ownerReference to include the MustGather CR.
+func (r *MustGatherReconciler) ensureObfuscateConfigMap(ctx context.Context, reqLogger logr.Logger, instance *mustgatherv1alpha1.MustGather) error {
```
Comment (swghosh) [inline controllers/mustgather/mustgather_controller.go]: why do we need this config map to be in the operator namespace to begin with? Since obfuscation is user-controlled behavior we should prefer having this config map in the same namespace where the CR is created (ditto like the PVC resource).

Comment (swghosh) [reply controllers/mustgather/mustgather_controller.go]: otherwise I guess breach:

> **Goals**
> 7. Allow users to provide custom obfuscation rules via a ConfigMap reference (obfuscationConfigRef) to override the default config

from the [EP](https://github.com/openshift/enhancements/blob/8c23d99d98c4ff13d419bee2dd82d543763c62a8/enhancements/support-log-gather/must-gather-bundle-obfuscation.md#goals), no?

File: controllers/mustgather/constant.go [hunk 5]
```diff
@@ -22,4 +22,18 @@ const (
...
+
+	// Obfuscation custom ConfigMap volume/mount paths.
+	obfuscateConfigVolumeName = "obfuscate-config"
+	obfuscateConfigMountPath  = "/etc/must-gather-clean/custom-config/config.yaml"
+	obfuscateConfigMapKey     = "config.yaml"
+
+	// obfuscateChownSuffix transfers gather output ownership to the upload container UID (65534).
+	// Captures the gather exit status first, runs chown (|| true so non-root images don't
+	// cause retries), then exits with the original status so gather failures propagate.
+	obfuscateChownSuffix = "gather_rc=$?; chown -R 65534:65534 /must-gather || true; exit $gather_rc"
```
Comment (swghosh) [inline controllers/mustgather/constant.go:38]: so if the previous command (gather) exited with a non-zero exit code this (chown execution) will fail immediately,
am I reading it right?

File: controllers/mustgather/template.go [hunk 6]
```diff
@@ -75,8 +76,64 @@ type GatherTimeFilter struct {
...
 
+func isObfuscateEnabled(obfuscate *v1alpha1.ObfuscateConfig) bool {
+	return obfuscate != nil && obfuscate.Enabled != nil && *obfuscate.Enabled
+}
+
+func shouldAppendObfuscateChown(obfuscate *v1alpha1.ObfuscateConfig) bool {
+	return isObfuscateEnabled(obfuscate) && obfuscate.Source == nil
+}
+
+type uploadSFTPParams struct {
```
Comment (swghosh) [inline controllers/mustgather/template.go]: let's re-use `SFTPSpec` from api/ ?

File: controllers/mustgather/template.go [hunk 7]
```diff
@@ -75,8 +76,64 @@ type GatherTimeFilter struct {
...
+	internalUser bool
+	secretRef    corev1.LocalObjectReference
+}
+
+func sftpUploadParams(mustGather v1alpha1.MustGather) *uploadSFTPParams {
+	if mustGather.Spec.UploadTarget == nil || mustGather.Spec.UploadTarget.Type != v1alpha1.UploadTypeSFTP {
+		return nil
+	}
+	s := mustGather.Spec.UploadTarget.SFTP
+	if s == nil || s.CaseID == "" || s.CaseManagementAccountSecretRef.Name == "" {
```
Comment (swghosh) [inline controllers/mustgather/template.go]: change this function into a `shouldSFTPUpload()` or `has...` instead and by re-using the api struct it should become cleaner, IMO.

File: controllers/mustgather/template.go [hunk 8]
```diff
@@ -268,17 +348,34 @@ func getGatherContainer(image string, audit bool, timeout time.Duration, storage
...
+		} else {
+			container.Command = command
+			if len(args) > 0 {
+				container.Args = args
+			}
+		}
 	} else {
+		gatherCmd := fmt.Sprintf(gatherCommand, math.Ceil(timeout.Seconds()), commandBinary)
+		if shouldAppendObfuscateChown(obfuscate) {
+			gatherCmd += "\n" + obfuscateChownSuffix
```
Comment (swghosh) [inline controllers/mustgather/template.go:350]: this is a little messy and occasionally can break your shell script, I feel - but for now non-blocking.

The way "\n"(s) over "bash -c" over Kubernetes container args, never been neat..

**Review by swghosh — COMMENTED** [review]
core logic seems to be accurate, a few nits most of them are all cosmetic ones - except the one concern about the obfuscate configMap namespace location which is blocking IMO.

Will review the e2e test cases.

File: test/e2e/must_gather_operator_test.go [hunk 9]
```diff
@@ -3603,6 +3617,615 @@ var _ = ginkgo.Describe("MustGather resource", ginkgo.Ordered, func() {
...
+						Enabled: func() *bool { b := true; return &b }(),
+						Source: &mustgatherv1alpha1.ObfuscateSourceConfig{
+							Claim: mustgatherv1alpha1.PersistentVolumeClaimReference{Name: mustGatherPVCName},
+						},
+					},
+				},
+			}
+			err := nonAdminClient.Create(testCtx, mg)
+			Expect(err).To(HaveOccurred(), "API should reject obfuscate.source with imageStreamRef")
+			Expect(err.Error()).To(ContainSubstring("obfuscate.source cannot be combined with imageStreamRef"))
```
Comment (swghosh) [inline test/e2e/must_gather_operator_test.go:3706]: we don't support obfuscation on custom must-gather images, yet?

File: test/e2e/must_gather_operator_test.go [hunk 10]
```diff
@@ -3603,6 +3617,615 @@ var _ = ginkgo.Describe("MustGather resource", ginkgo.Ordered, func() {
 			ginkgo.GinkgoWriter.Printf("Prefix1: %s\nPrefix2: %s\n", prefix1, prefix2)
 		})
 	})
+
+	// ─── Obfuscation Integration Tests ───────────────────────────────────
+
+	ginkgo.Context("Obfuscation CRD Validation Tests", func() {
```
Comment (swghosh) [inline test/e2e/must_gather_operator_test.go:3627]: non-blocking, in the future we should add envtest based CRD validation tests to house these CEL validation rules in a common place (outside the e2e).

File: test/e2e/must_gather_operator_test.go [hunk 11]
```diff
@@ -3603,6 +3617,615 @@ var _ = ginkgo.Describe("MustGather resource", ginkgo.Ordered, func() {
...
+					return apierrors.IsNotFound(adminClient.Get(testCtx,
+						client.ObjectKey{Name: configMapName, Namespace: ns.Name},
+						&corev1.ConfigMap{}))
+				}).WithTimeout(1*time.Minute).WithPolling(5*time.Second).Should(BeTrue(),
+					"Copied ConfigMap should be deleted after MustGather cleanup")
+			}
+		})
+	})
+
+	ginkgo.Context("Obfuscation Job template verification", func() {
```
Comment (swghosh) [inline test/e2e/must_gather_operator_test.go:4013]: we don't need this as long as we test similarly in template_test.go unit test cases.

File: test/e2e/must_gather_operator_test.go (hunk 9 above)
Comment (shivprakashmuley) [reply test/e2e/must_gather_operator_test.go:3706]: we can @swghosh. IMO `mg-clean` just works on a bundle file by file. I will just verify it manually once with some custom image and let you know.

File: controllers/mustgather/template.go [hunk 12]
```diff
@@ -122,59 +168,69 @@ func getJobTemplate(image string, operatorImage string, mustGather v1alpha1.Must
...
+		VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}},
+	}
+	if isObfuscateEnabled(obfuscate) && !hasObfuscateSource(obfuscate) &&
+		storage != nil && storage.Type == v1alpha1.StorageTypePersistentVolume {
+		uploadVolume.VolumeSource = corev1.VolumeSource{
+			PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+				ClaimName: storage.PersistentVolume.Claim.Name,
+			},
+		}
 	}
```
Comment (neha037) [inline controllers/mustgather/template.go]: You ask the operator to obfuscate an existing must-gather bundle already on a PVC (obfuscate.source), and you also set storage so the cleaned result should be saved. The API/docs/example all say that combo is how you “persist obfuscated output.”

What actually happens:
1. Input is wired correctly: the source PVC is mounted so obfuscation can read the old bundle.
2. Output is not: cleaned files are written to the upload volume at /must-gather-upload/cleaned.
3. That upload volume stays an emptyDir (temporary disk that dies with the pod) whenever obfuscate.source is set. The PVC from storage is never attached to that volume — the code only mounts storage onto the upload volume when source is not set.
4. With no upload credentials, the script finishes obfuscation and exits successfully. The job looks fine.
5. When the pod exits, emptyDir is wiped — the cleaned data is gone. Your storage PVC was never used for output.

So the CR validates and the job “succeeds,” but the thing you asked to keep (obfuscated output on PVC) never lands there.

File: controllers/mustgather/mustgather_controller.go [hunk 13]
```diff
@@ -674,3 +684,11 @@ func (r *MustGatherReconciler) cleanupTrustedCAConfigMap(ctx context.Context, re
 		"configMapName", r.TrustedCAConfigMap, "remainingNumOwners", len(updatedOwnerRefs))
 	return nil
 }
+
+func obfuscateConfigMapRefName(instance *mustgatherv1alpha1.MustGather) string {
```
Comment (swghosh) [inline controllers/mustgather/mustgather_controller.go]:
```suggestion
func getObfuscateConfigMapRefName(instance *mustgatherv1alpha1.MustGather) string {
```

File: controllers/mustgather/template.go [hunk 14]
```diff
@@ -75,8 +76,53 @@ type GatherTimeFilter struct {
...
+
+func hasSFTPUpload(mustGather v1alpha1.MustGather) bool {
+	if mustGather.Spec.UploadTarget == nil || mustGather.Spec.UploadTarget.Type != v1alpha1.UploadTypeSFTP {
+		return false
+	}
+	s := mustGather.Spec.UploadTarget.SFTP
+	return s != nil && s.CaseID != "" && s.CaseManagementAccountSecretRef.Name != ""
+}
+
+func sftpSpec(mustGather v1alpha1.MustGather) *v1alpha1.SFTPSpec {
```
Comment (swghosh) [inline controllers/mustgather/template.go]:
```suggestion
func getSftpSpec(mustGather v1alpha1.MustGather) *v1alpha1.SFTPSpec {
```

File: controllers/mustgather/template.go [hunk 15]
```diff
@@ -75,8 +76,53 @@ type GatherTimeFilter struct {
...
+		return nil
+	}
+	return mustGather.Spec.UploadTarget.SFTP
+}
+
+func shouldAddUploadContainer(mustGather v1alpha1.MustGather) bool {
+	return isObfuscateEnabled(mustGather.Spec.Obfuscate) || hasSFTPUpload(mustGather)
+}
+
+func obfuscateConfigMapName(obfuscate *v1alpha1.ObfuscateConfig) string {
```
Comment (swghosh) [inline controllers/mustgather/template.go]: redundant as of `obfuscateConfigMapRefName` in the other template.go file.

**Review by swghosh — APPROVED** [review]
Thank you @shivprakashmuley - this is a big feature :)

LGTM, few nits

@neha037 please add your /lgtm once you're able to re-review and the raised concern in https://github.com/openshift/must-gather-operator/pull/376#discussion_r3682652622 is resolved.

**Comment by shivprakashmuley** [comment]
/retitle [MG-157](https://redhat.atlassian.net/browse/MG-157): Integrate must-gather-clean in must-gather-operator.

[MG-157]: https://redhat.atlassian.net/browse/MG-157?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
